big_vision/trainers/proj/gsam/train.py

```diff
@@ -0,0 +1,370 @@
+  train_ds = input_pipeline.make_for_train(
+      dataset=config.dataset,
+      split=config.train_split,
+      batch_size=config.batch_size,
+      preprocess_fn=pp_builder.get_preprocess_fn(config.pp_train),
+      shuffle_buffer_size=config.get("shuffle_buffer_size"),
+      cache_raw=config.get("cache_raw", False),
+      data_dir=fillin(config.get("dataset_dir")))
+
+  n_prefetch = config.get("prefetch_to_device", 1)
+  train_iter = input_pipeline.start_input_pipeline(train_ds, n_prefetch)
+
+  ntrain_img = input_pipeline.get_num_examples(
+      config.dataset, config.train_split,
+      data_dir=fillin(config.get("dataset_dir")))
+  steps_per_epoch = ntrain_img / batch_size
+
+  if config.get("num_epochs"):
+    total_steps = int(config.num_epochs * steps_per_epoch)
+    assert not config.get("total_steps"), "Set either num_epochs or total_steps"
+  else:
+    total_steps = config.total_steps
+
+  info("Running for %d steps, that means %f epochs and %f steps per epoch",
+       total_steps, total_steps * batch_size / ntrain_img, steps_per_epoch)
+
+  write_note(f"Initializing {config.model_name} model...")
+  model_mod = importlib.import_module(f"big_vision.models.{config.model_name}")
+  model = model_mod.Model(
+      num_classes=config.num_classes, **config.get("model", {}))
+
+  # We want all parameters to be created in host RAM, not on any device, they'll
+  # be sent there later as needed, otherwise we already encountered two
+  # situations where we allocate them twice.
+  @partial(jax.jit, backend="cpu")
+  def init(rng):
+    shape = tuple(train_ds.element_spec["image"].shape[1:])
+    bs = config.batch_size // jax.device_count()
+    dummy_input = jnp.zeros((bs,) + shape, jnp.float32)
+    params = flax.core.unfreeze(model.init(rng, dummy_input))["params"]
+
+    # Set bias in the head to a low value, such that loss is small initially.
+    if "init_head_bias" in config:
+      params["head"]["bias"] = jnp.full_like(params["head"]["bias"],
+                                             config["init_head_bias"])
+
+    return params
+
+  rng, rng_init = jax.random.split(rng)
+  params_cpu = init(rng_init)
+
+  if jax.process_index() == 0:
+    num_params = sum(p.size for p in jax.tree_leaves(params_cpu))
+    parameter_overview.log_parameter_overview(params_cpu, msg="init params")
+    mw.measure("num_params", num_params)
+
+  write_note(f"Initializing {config.optax_name} optimizer...")
+  tx, sched_fns = bv_optax.make(config, params_cpu, sched_kw=dict(
+      global_batch_size=batch_size,
+      total_steps=total_steps,
+      steps_per_epoch=steps_per_epoch))
+
+  assert len(sched_fns) == 1, "Current GSAM only supports one global learning-rate."
+
+  # We jit this, such that the arrays are created on the CPU, not device[0].
+  opt_cpu = jax.jit(tx.init, backend="cpu")(params_cpu)
+  sched_fns_cpu = [jax.jit(sched_fn, backend="cpu") for sched_fn in sched_fns]
+
+  @partial(jax.pmap, axis_name="batch", donate_argnums=(0, 1))
```

Comment:
You need to add `, static_broadcasted_argnums=(5,))` here or this will not work: `step` is a scalar, so we need to tell `pmap` that, or it expects it to be replicated. So the final line should look like:

```
  @partial(jax.pmap, axis_name="batch", donate_argnums=(0, 1),
           static_broadcasted_argnums=(5,))
  def update_fn...
```

Reply:
wait no that's not what we should do, or it will recompile a new function every step :sweat_smile: Instead, we should indeed replicate the step we're passing, for example by passing `flax.jax_utils.replicate(step)` at call-site.

However, this is creating a synchronization point, blocks prefetching, and creates a transfer at each step. Instead, we should really use the step number which is already replicated inside the optimizer. I'll find out how exactly tomorrow.